### PR TITLE
build: bump rules_foreign_cc to e3f4b5e0 and switch to lld for .bazel…

### PR DESCRIPTION
**Pull request by htuch**
…rc sanitizers.

This allows for all --copt/--cxxopt/--linkopt to be plumbed to external
dependencies.

Risk level: Low
Testing: CI

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

**Comment by lizan** [comment]
@htuch ASAN/TSAN test failed, so the flags are propagated down to dependencies?

**Comment by htuch** [comment]
@mergeconflict is looking at the TSAN failure related to `evwatch`.

**Comment by mergeconflict** [comment]
> @mergeconflict is looking at the TSAN failure related to `evwatch`.

https://github.com/libevent/libevent/pull/802

**Comment by mergeconflict** [comment]
> > @mergeconflict is looking at the TSAN failure related to `evwatch`.
> 
> [libevent/libevent#802](https://github.com/libevent/libevent/pull/802)

#6637

File: .bazelrc [hunk 1]
```diff
@@ -14,6 +14,7 @@ build:asan --copt -fsanitize=address,undefined
 build:asan --linkopt -fsanitize=address,undefined
 build:asan --copt -fno-sanitize=vptr
 build:asan --linkopt -fno-sanitize=vptr
+build:asan --linkopt -fuse-ld=lld
```
Comment (lizan) [inline .bazelrc]: lld is already enabled in ci/build_setup.sh, though it is fine to have this here too.

File: source/extensions/filters/common/lua/lua.h [hunk 2]
```diff
@@ -11,6 +11,8 @@
 #include "common/common/c_smart_ptr.h"
 #include "common/common/logger.h"
 
+#define LUAJIT_ENABLE_GC64
```
Comment (lizan) [inline source/extensions/filters/common/lua/lua.h]: Is this good? I think you need define them when building LuaJIT, no?

Comment (htuch) [reply source/extensions/filters/common/lua/lua.h]: That was leftover from experiments. Turns out GC64 doesn't solve the alignment issues, so I'm just going to blacklist ASAN over the entire LuaJIT code base. I'm looking at the zlib issues that remain now.

File: .bazelrc (hunk 1 above)
Comment (htuch) [reply .bazelrc]: Yes, but when I build locally I also want `lld`.

File: source/extensions/filters/common/lua/lua.h (hunk 2 above)
Comment (htuch) [reply source/extensions/filters/common/lua/lua.h]: https://github.com/madler/zlib/issues/419

**Comment by htuch** [comment]
@lizan this is ready for final review, all tests should now pass.

File: bazel/foreign_cc/luajit.patch [hunk 3]
```diff
@@ -73,6 +73,13 @@ index 0000000..9c71271
 +    # fail on it.
 +    os.environ["LSAN_OPTIONS"] = "exitcode=0"
 +
++    # Blacklist LuaJIT from ASAN for now.
++    # TODO(htuch): Remove this when https://github.com/envoyproxy/envoy/issues/6084 is resolved.
++    if "ENVOY_CONFIG_ASAN" in os.environ:
++      os.environ["TARGET_CFLAGS"] += " -fsanitize-blacklist=%s/com_github_luajit_luajit/clang-asan-blacklist.txt" % os.environ["PWD"]
```
Comment (lizan) [inline bazel/foreign_cc/luajit.patch]: if we do `fun:*` blacklist, shall we just remove -fsanitize=address from TARGET_CFLAGS?

Comment (htuch) [reply bazel/foreign_cc/luajit.patch]: It's a useful hook to leave around to experiment with making this more precise, so I'd vote to keep it as is.

Comment (lizan) [reply bazel/foreign_cc/luajit.patch]: SGTM

**Pull request merged**